atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheImpl.java

```diff
@@ -0,0 +1,223 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.api.watch;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.palantir.atlasdb.keyvalue.api.watch.TimestampToVersionMap.CommitInfo;
+import com.palantir.atlasdb.transaction.api.TransactionLockWatchFailedException;
+import com.palantir.lock.LockDescriptor;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.watch.ClientLockWatchEventLog;
+import com.palantir.lock.watch.ClientLogEvents;
+import com.palantir.lock.watch.CommitUpdate;
+import com.palantir.lock.watch.IdentifiedVersion;
+import com.palantir.lock.watch.ImmutableInvalidateAll;
+import com.palantir.lock.watch.ImmutableInvalidateSome;
+import com.palantir.lock.watch.LockEvent;
+import com.palantir.lock.watch.LockWatchCreatedEvent;
+import com.palantir.lock.watch.LockWatchEvent;
+import com.palantir.lock.watch.LockWatchEventCache;
+import com.palantir.lock.watch.LockWatchStateUpdate;
+import com.palantir.lock.watch.NoOpLockWatchEventCache;
+import com.palantir.lock.watch.TransactionUpdate;
+import com.palantir.lock.watch.TransactionsLockWatchEvents;
+import com.palantir.lock.watch.UnlockEvent;
+import com.palantir.logsafe.Preconditions;
+
+/**
+ * This class should only be used through {@link ResilientLockWatchEventCache} as a proxy; failure to do so will result
+ * in concurrency issues and inconsistency in the cache state.
+ */
+public final class LockWatchEventCacheImpl implements LockWatchEventCache {
+    private final ClientLockWatchEventLog eventLog;
+    private final TimestampToVersionMap timestampMap;
+
+    public static LockWatchEventCache create() {
+        return ResilientLockWatchEventCache.newProxyInstance(
+                new LockWatchEventCacheImpl(ClientLockWatchEventLogImpl.create()), NoOpLockWatchEventCache.INSTANCE);
+    }
+
+    @VisibleForTesting
+    LockWatchEventCacheImpl(ClientLockWatchEventLog eventLog) {
+        this.eventLog = eventLog;
+        timestampMap = new TimestampToVersionMap();
+    }
+
+    @Override
+    public Optional<IdentifiedVersion> lastKnownVersion() {
+        return eventLog.getLatestKnownVersion();
+    }
+
+    @Override
+    public void processStartTransactionsUpdate(
+            Set<Long> startTimestamps,
+            LockWatchStateUpdate update) {
+        Optional<IdentifiedVersion> latestVersion = processEventLogUpdate(update);
+
+        latestVersion.ifPresent(
+                version -> startTimestamps.forEach(timestamp -> timestampMap.putStartVersion(timestamp, version)));
+
+        getEarliestVersion().ifPresent(eventLog::removeOldEntries);
```

Comment:
I've considered this before - the reason I went for it is that this will be called once per batch, as opposed to once per every transaction. But then again, this should be fast and I probably shouldn't micro-optimise at the cost of clarity, so I'll move it.

Reply:
Went back on this decision, as I think it's more correct in process on start: what if you are tracking events before you ever put a timestamp in? This way, you will clear all of those events out when you do a first put.